Simplify readme (#5217)
koreader/koreader: doc/Building_targets.md

This will allow you to build the *exact* same TCs used to build the nightlies, thanks to the magic of [crosstool-ng](https://github.com/crosstool-ng/crosstool-ng). These are also included precompiled in the Docker images for the respective targets.

**NOTE 3:** The vendor toolchain will be downloaded automatically by `./kodev release pocketbook`

### Additional packages

Some platforms will require additional packages:

#### for Android

Building for Android requires `openjdk-8-jdk` and `p7zip-full`.

For both Ubuntu and Debian, install the packages:

```
sudo apt-get install openjdk-8-jdk p7zip-full
```

#### for Debian

Building a debian package requires the `dpkg-deb` tool. It should be already installed if you're on a Debian/Ubuntu based distribution.

#### for Ubuntu Touch

Building for Ubuntu Touch requires the `click` package management tool.

Ubuntu users can install it with:

```
sudo apt-get install click
```

**NOTE**: The Ubuntu Touch build won't start anymore, and none of the currently active developers have any physical devices. Please visit [#4960](

> https://github.com/koreader/koreader/issues/4960) if you want to help.

Reviewer: We should probably also mention that builds still exist, so that if you just want to try to get the error logs you can get those without compiling anything.
```suggestion
https://github.com/koreader/koreader/issues/4960) if you want to help.
The Ubuntu Touch builds are therefore no longer published under releases on GitHub, but they are still available from [the nightly build server](http://build.koreader.rocks/download/nightly/).
```

# Building targets

KOReader is available for multiple platforms. Here are instructions to build installable packages for all these platforms.

These instructions are intended for a Linux OS. MacOS and Windows users are suggested to develop in a Linux VM.

## Prerequisites

> This instructions asume that you [have a development environment ready to run](Building.md) KOReader. If not then please install common prerequisites first.

Reviewer: There's a premade Docker image available for every supported target.

Author: I would need a bit of help with that since I was unaware and I never used Docker beyond a few tests.

Reviewer: You mean to mention specific commands in this file? That's possible too. I was basically just thinking of referencing the other README file in the virdevenv repo, which could use some more expanding. Currently more or less presumes you know Docker.

You need Git & Docker, or even just Docker will do. Assuming you want to interact with the koreader dir from your regular environment (text editor etc., but normally also git-related commands) you run:

```bash
docker run -v $(pwd)/koreader:/home/ko/koreader -it koreader/koappimage:latest bash
# Now you're in Bash in the Docker image, and commands like ./kodev build work out of the box
# The Docker image includes Git, hence why you don't need it on the host
# So besides the docker run part everything is the same as in the guide here.
# Docker gives you ready-made prerequisites.
```

Where "koappimage" is whatever target you're looking for. The appimage one will take care of all your regular desktop needs. Actually the baseimage can already do Debian packages and basic emulator; you'd need to run it on the host with SDL installed in that case. Only the AppImage one can run the emulator *from* Docker.

The `-v` flag mounts your local directory in the Docker container when it runs. This isn't required, but otherwise Docker tosses out all changes (i.e., downloads & compiled binaries) on exit. 

I hope that clarifies things a little bit.

### A toolchain for your target.

> Each target has its own architecture and you'll need to setup a proper cross-compile toolchain. Your GCC should be at least version 4.8

Reviewer: I just bumped that to 4.9 with https://github.com/koreader/koreader-base/pull/917
```suggestion
Each target has its own architecture and you'll need to setup a proper cross-compile toolchain. Your GCC should be at least version 4.9.
```

#### for Android

A compatible version of the Android NDK and SDK will be downloaded automatically by `./kodev release android` if no NDK or SDK is provided in environment variables. For that purpose you can use:

```
NDK=/ndk/location SDK=/sdk/location ./kodev release android
```

> If you want to use your own installed tools please make sure that you have the **NDKr15c** and the SDK for Android 9 (**api level 28**) already installed.

Reviewer: ```suggestion
If you want to use your own installed tools please make sure that you have the **NDKr15c** and the SDK for Android 9 (**API level 28**) already installed.
```

#### for embedded linux devices

Cross compile toolchains are available for Ubuntu users through these commands:

##### Kindle and Cervantes

```
sudo apt-get install gcc-arm-linux-gnueabi g++-arm-linux-gnueabi
```

##### Kobo and Ubuntu Touch

```
sudo apt-get install gcc-arm-linux-gnueabihf g++-arm-linux-gnueabihf
```

**NOTE 1:** The packages `pkg-config-arm-linux-gnueabihf` and `pkg-config-arm-linux-gnueabi` may

> block you from building. Remove them if you get the ld error

Reviewer: ```suggestion
block you from building. Remove them if you get the following ld error
```

```
/usr/lib/gcc-cross/arm-linux-gnueabihf/4.8/../../../../arm-linux-gnueabihf/bin/ld: cannot find -lglib-2.0
```

**NOTE 2:** In the specific case of Cervantes, Kindle & Kobo targets, while we make some effort to support these Linaro/Ubuntu TCs,
they do *not* exactly target the proper devices. While your build will go fine, this may lead to runtime failure.
As time goes by, and/or the more bleeding-edge your distro is, the greater the risk for mismatch gets.
Thankfully, we have a distribution-agnostic solution for you: [koxtoolchain](https://github.com/koreader/koxtoolchain)!

> This will allow you to build the *exact* same TCs used to build the nightlies, thanks to the magic of [crosstool-ng](https://github.com/crosstool-ng/crosstool-ng).

Reviewer: ```suggestion
This will allow you to build the *exact* same TCs used to build the nightlies, thanks to the magic of [crosstool-ng](https://github.com/crosstool-ng/crosstool-ng). These are also included precompiled in the Docker images for the respective targets.
```

**NOTE 3:** The vendor toolchain will be downloaded automatically by `./kodev release pocketbook`

### Additional packages

> Some platforms will require aditional packages:

Reviewer: ```suggestion
Some platforms will require additional packages:
```

#### for Android

Building for Android requires `openjdk-8-jdk` and `p7zip-full`.

> Users of Debian Jessie first need to configure the `backports` repository:

Reviewer: Should we remove this? I added it when Jessie was the stable, that's Stretch now. Then again, it's still on LTS for a while longer.

```
sudo echo "deb http://ftp.debian.org/debian jessie-backports main" > /etc/apt/sources.list.d/backports.list
sudo apt-get update
```

For both Ubuntu and Debian, install the packages:

```
sudo apt-get install openjdk-8-jdk p7zip-full
```

Users on Debian finally need to remove JRE version 7:

```
sudo apt-get remove openjdk-7-jre-headless
```

#### for Debian

Building a debian package requires the `dpkg-deb` tool. It should be already installed if you're on a Debian/Ubuntu based distribution.

#### for Ubuntu Touch

> 

Reviewer: Our Ubuntu Touch support is in limbo. It may not be overly hard to get working again as a proof of concept provided a dev had a physical device (no emulator), but in the longer term a different build process would need to be used since this one won't work past Ubuntu 16.04.

So we could either remove this section or add a bold warning of some sorts, along these lines:
```suggestion
The Ubuntu Touch build won't start anymore, and none of the currently active developers have any physical devices. Please visit [#4960](
https://github.com/koreader/koreader/issues/4960) if you want to help.
```

Author: Added a note

Building for Ubuntu Touch requires the `click` package management tool.

Ubuntu users can install it with:

```
sudo apt-get install click
```

## Building 

> You may checkout our [nightlybuild script][nb-script] to see how to build a

Reviewer: ```suggestion
You can check out our [nightlybuild script][nb-script] to see how to build a
```
